p2pserver/protocols/msg_handler.go

```diff
@@ -39,35 +40,60 @@ import (
 	"github.com/ontio/ontology/p2pserver/protocols/heatbeat"
 	"github.com/ontio/ontology/p2pserver/protocols/recent_peers"
 	"github.com/ontio/ontology/p2pserver/protocols/reconnect"
+	"github.com/ontio/ontology/p2pserver/protocols/subnet"
+	"github.com/ontio/ontology/p2pserver/protocols/utils"
 )
 
 //respCache cache for some response data
-var respCache *lru.ARCCache
+var respCache, _ = lru.NewARC(msgCommon.MAX_RESP_CACHE_SIZE)
 
 //Store txHash, using for rejecting duplicate tx
 // thread safe
 var txCache, _ = lru.NewARC(msgCommon.MAX_TX_CACHE_SIZE)
 
 type MsgHandler struct {
+	seeds                    *utils.HostsResolver
 	blockSync                *block_sync.BlockSyncMgr
 	reconnect                *reconnect.ReconnectService
 	discovery                *discovery.Discovery
 	heatBeat                 *heatbeat.HeartBeat
 	bootstrap                *bootstrap.BootstrapService
 	persistRecentPeerService *recent_peers.PersistRecentPeerService
+	subnet                   *subnet.SubNet
 	ledger                   *ledger.Ledger
+	acct                     *account.Account // nil if conenesus is not enabled
+	staticReserveFilter      p2p.AddressFilter
 }
 
-func NewMsgHandler(ld *ledger.Ledger) *MsgHandler {
-	return &MsgHandler{ledger: ld}
+func NewMsgHandler(acct *account.Account, staticReserveFilter p2p.AddressFilter, ld *ledger.Ledger, logger msgCommon.Logger) *MsgHandler {
+	gov := utils.NewGovNodeResolver(ld)
+	seedsList := config.DefConfig.Genesis.SeedList
+	seeds, invalid := utils.NewHostsResolver(seedsList)
+	if invalid != nil {
+		panic(fmt.Errorf("invalid seed list； %v", invalid))
+	}
+	subNet := subnet.NewSubNet(acct, seeds, gov, logger)
+	return &MsgHandler{ledger: ld, seeds: seeds, subnet: subNet, acct: acct, staticReserveFilter: staticReserveFilter}
+}
+
+func (self *MsgHandler) GetReservedAddrFilter(staticFilterEnabled bool) p2p.AddressFilter {
+	return self.subnet.GetReservedAddrFilter(staticFilterEnabled)
+}
+
+func (self *MsgHandler) GetMaskAddrFilter() p2p.AddressFilter {
+	return self.subnet.GetMaskAddrFilter()
+}
+
+func (self *MsgHandler) GetSubnetMembersInfo() []msgCommon.SubnetMemberInfo {
+	return self.subnet.GetMembersInfo()
 }
 
 func (self *MsgHandler) start(net p2p.P2P) {
 	self.blockSync = block_sync.NewBlockSyncMgr(net, self.ledger)
-	self.reconnect = reconnect.NewReconectService(net)
-	self.discovery = discovery.NewDiscovery(net, config.DefConfig.P2PNode.ReservedCfg.MaskPeers, 0)
-	seeds := config.DefConfig.Genesis.SeedList
-	self.bootstrap = bootstrap.NewBootstrapService(net, seeds)
+	self.reconnect = reconnect.NewReconectService(net, self.staticReserveFilter)
+	maskFilter := self.subnet.GetMaskAddrFilter()
+	self.discovery = discovery.NewDiscovery(net, config.DefConfig.P2PNode.ReservedCfg.MaskPeers, maskFilter, 0)
+	self.bootstrap = bootstrap.NewBootstrapService(net, self.seeds)
```

Comment:
suggest to add reserved peers in peers.rsv to bootstrap, even --reserve-only is not configured.

So, we can still use peers.rsv to maintain subnet connections manually.